coreax/solvers/recombination.py

```diff
@@ -0,0 +1,570 @@
+    expressed as :math:`\mathcal{O}(m^3)`, using the same arguments as used in
+    :class:`CaratheodoryRecombination`.
+
+    ..note::
+        As the ratio of :math:`n / m` grows, the constant factor for the time complexity
+        of :class:`TreeRecombination` increases at a logarithmic rate, rather than at a
+        quadratic rate for plain :class:`CaratheodoryRecombination`. Hence, in general,
+        we would expect :class:`TreeRecombination` to be the more efficient choice for
+        all but the smallest values of :math:`n / m`.
+
+    :param test_functions: the map :math:`\Phi^\prime = \{ \phi_1, \dots, \phi_{M-1} \}`
+        where each :math:`\phi_i \colon \Omega \to \mathbb{R}` represents a linearly
+        independent test-function; a value of `None` implies the identity function
+        (necessarily assuming :math:`\Omega \subseteq \mathbb{R}^{M-1}`)
+    :param mode: 'implicit-explicit' explicitly removes :math:`n - m` points, yielding
+        a coreset of size :math:`m`, with :math:`m - m^\prime` zero-weighted (implicitly
+        removed) points; 'implicit' explicitly removes no points, yielding a coreset of
+        size :math:`n` with :math:`n - m^\prime` zero-weighted (implicitly removed)
+        points; 'explicit' explicitly removes :math:`n - m^\prime` points, yielding a
+        coreset of size :math:`m^\prime`, but unlike the other methods is not JIT
+        compatible as the coreset size :math:`m^\prime` is unknown at compile time.
+    :param rcond: a relative condition number; any singular value :math:`s` below the
+        threshold :math:`\text{rcond} * \text{max}(s)` is treated as equal to zero; if
+        :code:`rcond is None`, it defaults to `floating point eps * max(n, d)`
+    :param tree_reduction_factor: The factor by which each tree reduction step reduces
+        the number of non-zero points; the remaining number of non-zero nodes, after
+        performing recombination, is equal to `n_nodes / tree_reduction_factor`;
+    """
+
+    rcond: Union[float, None] = None
+    tree_reduction_factor: int = 2
+
+    @override
+    def reduce(
+        self, dataset: Data, solver_state: None = None
+    ) -> tuple[Coresubset, None]:
+        nodes, weights = jtu.tree_leaves(dataset.normalize(preserve_zeros=True))
+        # Push the nodes forward through the test-functions \Phi^\prime.
+        push_forward_nodes = _push_forward(nodes, self.test_functions, augment=False)
+        n, m = push_forward_nodes.shape
+        # We don't apply the affine-augmentation test-function \phi_1 here, instead
+        # deferring it to `CaratheodoryRecombination.reduce`. Thus, we have to manually
+        # correct the value for `m`.
+        padding, count, depth = _prepare_tree(n, m + 1, self.tree_reduction_factor)
+        car_recomb_solver = CaratheodoryRecombination(rcond=self.rcond, mode="implicit")
+
+        def _tree_reduce(_, state):
```

Comment:
What is the `_` parameter for here?

Reply:
This function is passed as the callable for `jax.lax.fori_loop`, which requires the function call signature to be of form `f(loop_iteration_count, carry_state)`. However, because we don't make use of the `loop_iteration_count` anywhere within the function we can name the parameter as `_` in the call signature. The use of `_` specifically is to suppres the `W0613 unused-argument` linting error that would otherwise be thrown by pylint. If you would prefer, we could alternatively do something like:

```python
def _tree_reduce(i, state):
    del i
    ...
```

Reply:
Thanks for the explanation, makes sense! Fine as is I think